drivers/clock_control/nrf_clock_calibration.c

```diff
@@ -0,0 +1,273 @@
+ * immediately after stop. Redefined events to avoid ifdefs in the code,
+ * CTSTOPPED interrupt handling will be removed during compilation.
+ */
+#ifndef CLOCK_EVENTS_CTSTOPPED_EVENTS_CTSTOPPED_Msk
+#define NRF_CLOCK_EVENT_CTSTOPPED 0
+#endif
+
+#ifndef CLOCK_INTENSET_CTSTOPPED_Msk
+#define NRF_CLOCK_INT_CTSTOPPED_MASK 0
+#endif
+
+#define TEMP_SENSOR_NAME \
+	COND_CODE_1(CONFIG_TEMP_NRF5, (CONFIG_TEMP_NRF5_NAME), (NULL))
+
+/* Calibration state enum */
+enum nrf_cal_state {
+	CAL_OFF,
+	CAL_IDLE,	/* Calibration timer active, waiting for expiration. */
+	CAL_HFCLK_REQ,	/* HFCLK XTAL requested. */
+	CAL_TEMP_REQ,	/* Temperature measurement requested. */
+	CAL_ACTIVE,	/* Ongoing calibration. */
+	CAL_ACTIVE_OFF /* Ongoing calibration, off requested. */
+};
+
+static enum nrf_cal_state cal_state; /* Calibration state. */
+static s16_t prev_temperature; /* Previous temperature measurement. */
+static u8_t calib_skip_cnt; /* Counting down skipped calibrations. */
+
+/* Callback called on hfclk started. */
+static void cal_hf_on_callback(struct device *dev, void *user_data);
+static struct clock_control_async_data cal_hf_on_data = {
+	.cb = cal_hf_on_callback
+};
+
+static struct device *hfclk_dev; /* Handler to hfclk device. */
+static struct device *temp_sensor; /* Handler to temperature sensor device. */
+
+static void measure_temperature(struct k_work *work);
+static K_WORK_DEFINE(temp_measure_work, measure_temperature);
+
+static bool clock_event_check_and_clean(u32_t evt, u32_t intmask)
+{
+	bool ret = nrf_clock_event_check(evt) &&
+			nrf_clock_int_enable_check(intmask);
+
+	if (ret) {
+		nrf_clock_event_clear(evt);
+	}
+
+	return ret;
+}
+
+bool nrf_clock_control_calibration_start(struct device *dev)
+{
+	bool ret;
+	int key = irq_lock();
+
+	if (cal_state != CAL_ACTIVE_OFF) {
+		nrf_clock_task_trigger(NRF_CLOCK_TASK_CTSTART);
+		ret = true;
+	} else {
+		ret = false;
+	}
+
+	cal_state = CAL_IDLE;
+
+	irq_unlock(key);
+
+	calib_skip_cnt = 0;
+
+	return ret;
+}
+
+bool nrf_clock_control_calibration_stop(struct device *dev)
+{
+	int key;
+	bool ret = true;
+
+	key = irq_lock();
+
+	nrf_clock_task_trigger(NRF_CLOCK_TASK_CTSTOP);
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_CTTO);
+
+	/* If calibration is active then pend until completed.
+	 * Currently (and most likely in the future), LFCLK is never stopped so
+	 * it is not an issue.
+	 */
+	if (cal_state == CAL_ACTIVE) {
+		cal_state = CAL_ACTIVE_OFF;
+		ret = false;
+		LOG_INF("Stop when calibration is active");
+	} else {
+		cal_state = CAL_OFF;
+	}
+
+	irq_unlock(key);
+
+	return ret;
+}
+
+void nrf_clock_control_calibration_init(struct device *dev)
+{
+	/* Errata 36: After watchdog timeout reset, CPU lockup reset, soft
+	 * reset, or pin reset EVENTS_DONE and EVENTS_CTTO are not reset.
+	 */
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_DONE);
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_CTTO);
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_CTSTOPPED);
```

Comment:
This should be inside `if (NRF_CLOCK_EVENT_CTSTOPPED) {}`.
It's not that harmful right not, as it clears `HFCLKSTARTED` if `CTSTOPPED` is not present, but...

Reply:
it's actually not needed. I've added it because i had some issues which turns out to be not related to that. forgot to remove it then. Removed now.